lightning/src/ln/channelmanager.rs

```diff
@@ -3321,6 +3367,104 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			self.finish_force_close_channel(failure);
 		}
 	}
+
+	/// Gets a payment secret and payment hash for use in an invoice given to a third party wishing
+	/// to pay us.
```

Comment:
I wonder if these two methods are trait-worthy. It may have (1) a single method for creating an inbound payment and generating/storing the preimage as it sees fit (rather than two methods) and (2) a method for retrieving this data including the preimage.

Or is there a reason why `PaymentReceived` should not contain a preimage that was generated elsewhere?

Reply:
Hmm, what exactly did you have in mind? A trait that `ChannelManager` implements? Or a trait for fetching payment preimages against a user implementation of some trait (presumably with some default implementation)?

In general I don't mind doing this via some external trait, but I do worry that we'd be adding a chunk of complexity to make users serialize yet another object that stores payment preimages (which is already the case, but I figured we could drop that requirement).

Reply:
Decided to keep the interface as-is for now after offline discussion on Slack. For posterity: https://lightningdevkit.slack.com/archives/CTBLT3CAU/p1619364792017100